### DISCUSSION TITLE
rehype-autolink-headings adds className to all anchor tags even without headings

### DISCUSSION THREAD

**Discussion by AnweshGangula** [rehypejs/rehype#86] [Q&A] [answered]
I'm working on a [Next.JS blog][1] (*referring to the [following article by LeeRob][2] - link to [source code][3]*) which uses MDX-Bundler  to convert MDX blog posts to HTML. I'm using the `rehype-autolink-headings` along with `rehype-slug` plugin to add clickable link on the headings of the post. 

The issue is that the `rehype-autolink-headings` plugin is adding the `className` to all anchor tags, even those which are not inside heading tags. Below is the configuration i'm using:

```js
xdmOptions(options) {
	options.rehypePlugins = [
		...(options.rehypePlugins ?? []),
		rehypeSlug,
		rehypeCodeTitles,
		rehypePrism,
		[
			rehypeAutolinkHeadings,
			{
				properties: {
					className: ['anchor']
				}
			}
		]
	];

	return options;
},
```

Below is the screenshot of the HTML from my blog.

[![Screenshot of HTML from my blog][4]][4]

  [1]: https://nextjstestag.azurewebsites.net/blog/sample%20markdown
  [2]: https://leerob.io/blog/mdx#mdx-plugins
  [3]: https://github.com/AnweshGangula/Nextjs-Blog/tree/91a589f96cee0a0b44db8be9b934e47e1541f652
  [4]: https://i.stack.imgur.com/S03R8.png

**Comment by ChristianMurphy** [comment]
@AnweshGangula have your reached out to the author of the blog post?
And could you share a runnable example of the issue you are seeing? https://next.new is a quick way to generate a sandbox.

`rehype-slug` only visits headings https://github.com/rehypejs/rehype-slug/blob/22d2ab6d11faa250052a341e21c566705fc72e73/index.js#L22-L23
as does `rehype-autolink-headings` https://github.com/rehypejs/rehype-autolink-headings/blob/2a6b77f80289acd84ab85644bf185e81832b5659/lib/index.js#L80-L84

I strongly suspect the issue is coming from another plugin or something the bundler is doing, but it's hard to say without seeing the exact code you are running and the exact versions of dependencies used with it.

**Reply by AnweshGangula** [reply]
I'm having some challenges in creating the [blog in Stackblitz](https://stackblitz.com/edit/nextjs-eedmnm) (_`mdx.js` is not working as exxpected_), but the source is in the following [GitHub repository](https://github.com/AnweshGangula/Nextjs-Blog).

**Reply by ChristianMurphy** [reply]
CodeSandbox is another alternative.
For example here is a sandbox which uses the official `@mdx-js/loader` (version 2 rc2), with `next` 12, `rehype-slug`, and `rehype-autolink-headings`. https://codesandbox.io/s/next-mdx-and-rehype-autolink-headings-6eip8?file=/next.config.mjs
Notice that the issue doesn't appear to happen with these packages alone, this strongly suggests another package is causing the issue.

**Comment by wooorm** [comment]
This is 99.99% not a problem with the plugins rehype-slug or rehype-autolink-headings.

**Reply by AnweshGangula** [reply]
I'm using MDX-Bundler which is a new package. Could that be causing this issue? 
I have the source code is in the following [GitHub repository](https://github.com/AnweshGangula/Nextjs-Blog).

**Reply by wooorm** [reply]
It could be many things.

Linking to a whole repo is too much, I don't have time to delve into a complete project. You can try and reduce your problem, make it as small as possible, to figure out what the problem is.

**Reply by AnweshGangula** [reply]
This project is actually very basic & built using the default Nextjs starter blog. The only custom components here are [`[slug].js`](https://github.com/AnweshGangula/Nextjs-Blog/blob/91a589f96cee0a0b44db8be9b934e47e1541f652/pages/blog/%5Bslug%5D.js) and [`mdx.js`](https://github.com/AnweshGangula/Nextjs-Blog/blob/91a589f96cee0a0b44db8be9b934e47e1541f652/utils/mdx.js). 

And the [`index.js`](https://github.com/AnweshGangula/Nextjs-Blog/blob/91a589f96cee0a0b44db8be9b934e47e1541f652/pages/index.js) is just customized to list all [blog posts](https://github.com/AnweshGangula/Nextjs-Blog/tree/91a589f96cee0a0b44db8be9b934e47e1541f652/data/posts).

I hope this helps.

**Comment by AnweshGangula** [comment, marked as answer]
Thank you @wooorm  @ChristianMurphy. 

Turns out that the issue was due to the [component-substitution](https://github.com/kentcdodds/mdx-bundler#component-substitution) that I have in my [`[slug.js]`](https://github.com/AnweshGangula/Nextjs-Blog/blob/91a589f96cee0a0b44db8be9b934e47e1541f652/pages/blog/%5Bslug%5D.js) file.
This was over-writing all the `classnames` and replacing them with the substitute component all anchor tags.

I'm really new to Next.js, so I was not able to reduce the problem to a specific file until now. Really Appreciate you help. 🙌🙏
